pkg/validator/runner.go

```diff
@@ -0,0 +1,222 @@
+package validator
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/mt-sre/addon-metadata-operator/pkg/types"
+	"github.com/mt-sre/addon-metadata-operator/pkg/utils"
+)
+
+var initializers []Initializer
+
+// Register queues an "Initializer" which the "Runner"
+// will invoke upon startup.
+func Register(init Initializer) {
+	initializers = append(initializers, init)
+}
+
+// Initializer is a function which will initialize a Validator
+// with dependencies. An error is returned if the Validator
+// cannot be initalized properly.
+type Initializer func(Dependencies) (Validator, error)
+
+// Dependencies abstracts common dependencies for Validators.
+type Dependencies struct {
+	Logger    logr.Logger
+	OCMClient OCMClient
+}
+
+// NewRunner returns a Runner configured with a variadic
+// slice of options or an error if an issue occurs.
+func NewRunner(opts ...RunnerOption) (*Runner, error) {
+	var cfg RunnerConfig
+
+	cfg.Option(opts...)
+
+	if err := cfg.Default(); err != nil {
+		return nil, err
+	}
+
+	deps := Dependencies{
+		Logger:    cfg.Logger,
+		OCMClient: cfg.OCMClient,
+	}
+
+	entries := make(map[Code]validatorEntry)
+
+	for _, init := range cfg.Initializers {
+		val, err := init(deps)
+		if err != nil {
+			return nil, err
+		}
+
+		if existing, ok := entries[val.Code()]; ok {
+			return nil, fmt.Errorf(
+				"code '%d' is already registered for validator '%s'",
+				val.Code(),
+				existing.Name(),
+			)
+		}
+
+		entries[val.Code()] = validatorEntry{
+			Validator: val,
+		}
+	}
+
+	return &Runner{
+		RunnerConfig: cfg,
+		entries:      entries,
+	}, nil
+}
+
+type Runner struct {
+	RunnerConfig
+	entries map[Code]validatorEntry
+}
+
+func (r *Runner) Run(ctx context.Context, mb types.MetaBundle, filters ...Filter) <-chan Result {
+	resultCh := make(chan Result)
+
+	var wg sync.WaitGroup
+
+	vals := r.GetValidators(filters...)
+
+	wg.Add(len(vals))
+
+	for _, val := range vals {
+		go func(v Validator) {
+			defer wg.Done()
+
+			select {
+			case <-ctx.Done():
+			case resultCh <- v.Run(ctx, mb):
+			}
+		}(val)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resultCh)
+	}()
+
+	return resultCh
+}
```

Comment:
Fire fire :+1: 

We did not use any multithreading before if I remember correctly.